File: frontend/src/lib/api/schemas/storyboards.ts
import { z } from "zod";
import { StoryboardStatus } from "@/types/enums";

const storyboardStatusSchema = z.enum([
  StoryboardStatus.DRAFT,
  StoryboardStatus.READY,
  StoryboardStatus.APPROVED,
]);

export const storyboardReadSchema = z.object({
  id: z.string().uuid(),
  project_id: z.string().uuid(),
  script_id: z.string().uuid(),
  status: storyboardStatusSchema,
});
export type StoryboardRead = z.infer<typeof storyboardReadSchema>;

export const sceneReadSchema = z.object({
  id: z.string().uuid(),
  storyboard_id: z.string().uuid(),
  sequence_index: z.number(),
  description: z.string(),
  duration_seconds: z.string(),
  image_prompt: z.string().nullable(),
  video_prompt: z.string().nullable(),
  voice_line: z.string().nullable(),
});
export type SceneRead = z.infer<typeof sceneReadSchema>;

export type StoryboardCreateRequest = {
  script_id: string;
};

export type StoryboardStatusUpdateRequest = {
  status: StoryboardStatus;
};

export type SceneCreateRequest = {
  sequence_index: number;
  description: string;
  duration_seconds: string;
  image_prompt?: string | null;
  video_prompt?: string | null;
  voice_line?: string | null;
};

File: backend/src/ytforge/interfaces/api/schemas/storyboards.py
from __future__ import annotations

import uuid
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from ytforge.domain.enums import StoryboardStatus


class StoryboardCreateRequest(BaseModel):
    script_id: uuid.UUID


class StoryboardStatusUpdateRequest(BaseModel):
    status: StoryboardStatus


class StoryboardRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    script_id: uuid.UUID
    status: StoryboardStatus


class SceneCreateRequest(BaseModel):
    sequence_index: int = Field(ge=0)
    description: str = Field(min_length=1)
    duration_seconds: Decimal = Field(gt=0)
    image_prompt: str | None = None
    video_prompt: str | None = None
    voice_line: str | None = None


class SceneRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    storyboard_id: uuid.UUID
    sequence_index: int
    description: str
    duration_seconds: Decimal
    image_prompt: str | None
    video_prompt: str | None
    voice_line: str | None


class SceneReorderRequest(BaseModel):
    ordered_scene_ids: list[uuid.UUID]
